packages/react/src/components/Tree/Tree.tsx

```diff
@@ -0,0 +1,334 @@
+  level: number
+  parent: ShorthandValue<TreeItemProps>
+  siblings: ShorthandCollection<TreeItemProps>
+}
+
+export interface TreeState {
+  activeItemIds: string[]
+  itemsForRender: { [s: string]: TreeItemForRenderProps }
+}
+
+class Tree extends AutoControlledComponent<WithAsProp<TreeProps>, TreeState> {
+  static create: Function
+
+  static displayName = 'Tree'
+
+  static className = 'ui-tree'
+
+  static slotClassNames: TreeSlotClassNames = {
+    item: `${Tree.className}__item`,
+  }
+
+  static propTypes = {
+    ...commonPropTypes.createCommon({
+      content: false,
+    }),
+    activeItemIds: customPropTypes.collectionShorthand,
+    defaultActiveItemIds: customPropTypes.collectionShorthand,
+    exclusive: PropTypes.bool,
+    items: customPropTypes.collectionShorthand,
+    renderItemTitle: PropTypes.func,
+    rtlAttributes: PropTypes.func,
+  }
+
+  static defaultProps = {
+    as: 'div',
+    accessibility: treeBehavior,
+  }
+
+  static autoControlledProps = ['activeItemIds']
+
+  static isSubtree(item: TreeItemProps | ShorthandValue<TreeItemProps>): boolean {
+    return !!item['items'] && item['items'].length > 0
+  }
+
+  static getItemsForRender = _.memoize((itemsFromProps: ShorthandCollection<TreeItemProps>) => {
+    const itemsForRenderGenerator = (
+      items = itemsFromProps,
+      level = 1,
+      parent?: ShorthandValue<TreeItemProps>,
+    ) => {
+      return _.reduce(
+        items,
+        (acc: Object, item: ShorthandValue<TreeItemProps>, index: number) => {
+          const id = item['id']
+          const isSubtree = Tree.isSubtree(item)
+
+          acc[id] = {
+            elementRef: React.createRef<HTMLElement>(),
+            level,
+            index,
+            parent,
+            siblings: items.filter(currentItem => currentItem !== item),
+          }
+
+          return {
+            ...acc,
+            ...(isSubtree ? itemsForRenderGenerator(item['items'], level + 1, item) : {}),
+          }
+        },
+        {},
+      )
+    }
+
+    return itemsForRenderGenerator(itemsFromProps)
+  })
+
+  static getAutoControlledStateFromProps(nextProps: TreeProps, prevState: TreeState) {
+    const itemsForRender = Tree.getItemsForRender(nextProps.items)
+
+    return {
+      itemsForRender,
+    }
+  }
+
+  getInitialAutoControlledState() {
+    return { activeItemIds: [] }
+  }
+
+  treeRef = React.createRef<HTMLElement>()
+
+  handleTreeItemOverrides = (predefinedProps: TreeItemProps) => ({
+    onTitleClick: (
+      e: React.SyntheticEvent,
+      treeItemProps: TreeItemProps,
+      predefinedProps: TreeItemProps,
+    ) => {
+      const { activeItemIds } = this.state
+      const { id, siblings } = treeItemProps
+      const { exclusive } = this.props
+
+      if (!Tree.isSubtree(treeItemProps)) {
+        return
+      }
+
+      const indexOfActiveItem = activeItemIds.indexOf(id)
+
+      if (indexOfActiveItem > -1) {
+        activeItemIds.splice(indexOfActiveItem, 1)
```

Comment:
This mutates original array 😥

Reply:
it's our array in our state. why should we care? is it the same reference as the potentially controlled one from props?

Reply:
It can cause the same issues as are described there:
https://daveceddia.com/why-not-modify-react-state-directly/

Because this code actually does:
```tsx
this.state.activeItemIds.splice(indexOfActiveItem, 1)
```